### PR TITLE
Fixed bug in parsing {..} regex

### LINKED ISSUE #38498: `{0..10}` expansion in external file names is incorrect.
It is interpreted as an "aligned" expansion producing something like 00, 01, 02... 10 instead of 0, 1, 2... 10.

https://pastila.nl/?09014b44/8155336bcfbf55d813814753465a1adf

### PR DESCRIPTION
**Pull request by HeenaBansal2009**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixes #38498
Current Implementation  is similar to what shell does mentiond  by @rschu1ze [here](https://github.com/ClickHouse/ClickHouse/pull/38502#issuecomment-1169057723)

### CODE REVIEW

**Comment by qoega** [comment]
Can you add a test for this case?

**Comment by HeenaBansal2009** [comment]
> Can you add a test for this case?



> > Can you add a test for this case?
> 
> --> Common/tests/gtest_makeRegexpPatternFromGlobs.cpp.

Sure, I'll add

**Comment by HeenaBansal2009** [comment]
@rschu1ze  @qoega ,
According to current implementation , 
{000..999} returns f(000|001|002| ....... |999) 
and 
{000..009} returns f(0|1|2|3.. |9)  ---> shouldn't it be f(000|001|002|003 ...009)
and 
{0..20} and {00..20} returns f(00|01|02|03|04|........ 19|20)"

So What is expected in case of {0..20}   and {00..20}, should  both return f(0|1|2|3|4 .. 20) OR  f(0|1|2|3|4 .. |20) and f(00|01|02..|20) respectively.

**Comment by rschu1ze** [comment]
To minimize confusion, we ideally provide 100% the same semantics as the shell does.

```
$ echo {000..999}
000 001 002 003 ... 9999

$ echo {000..009}
000 001 002 003 004 005 006 007 008 009
# yes you are right, ClickHouse is currently wrong in this case

$ echo {0..20}
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20
# again, ClickHouse is wrong here

$ echo  {00..20}
00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20
```

**Comment by alexey-milovidov** [comment]
A functional test with clickhouse-local, temporary files and a file table function is possible to make.

File: src/Common/parseGlobs.cpp [hunk 1]
```diff
@@ -51,7 +51,15 @@ std::string makeRegexpPatternFromGlobs(const std::string & initial_str_with_glob
             ReadBufferFromString buf_range(buffer);
             buf_range >> range_begin >> point >> point >> range_end;
             bool leading_zeros = buffer[0] == '0';
-            size_t num_len = std::to_string(range_end).size();
+
```
Comment (rschu1ze) [inline src/Common/parseGlobs.cpp:54]: Totally not related to your PR, but in l. 35, would you like to replace `{}*,` in the comment by `"{", "}", "*" and ","`?

File: src/Common/parseGlobs.cpp [hunk 2]
```diff
@@ -51,7 +51,15 @@ std::string makeRegexpPatternFromGlobs(const std::string & initial_str_with_glob
...
+            size_t range_begin_width = buffer.find('.');
+            size_t range_end_width = buffer.size() - buffer.find_last_of('.') - 1;
+
+            size_t num_len = 0;
+            if (range_begin_width == 1 && leading_zeros)
+                num_len = 1;       /// Scenarios {0..10} {0..999}
+            else
+                num_len = range_begin_width < range_end_width ? range_end_width : range_begin_width;       /// Scenarios {00..99} {00..099}
```
Comment (rschu1ze) [inline src/Common/parseGlobs.cpp:62]: `num_len = std::max(range_begin_width, range_end_width)`

File: src/Common/parseGlobs.cpp [hunk 3]
```diff
@@ -51,7 +51,15 @@ std::string makeRegexpPatternFromGlobs(const std::string & initial_str_with_glob
...
             buf_range >> range_begin >> point >> point >> range_end;
             bool leading_zeros = buffer[0] == '0';
-            size_t num_len = std::to_string(range_end).size();
+
+            size_t range_begin_width = buffer.find('.');
+            size_t range_end_width = buffer.size() - buffer.find_last_of('.') - 1;
+
+            size_t num_len = 0;
+            if (range_begin_width == 1 && leading_zeros)
+                num_len = 1;       /// Scenarios {0..10} {0..999}
```
Comment (rschu1ze) [inline src/Common/parseGlobs.cpp:60]: Maybe add a "special case: " at the beginning of the comment?

File: src/Common/parseGlobs.cpp [hunk 4]
```diff
@@ -51,7 +51,15 @@ std::string makeRegexpPatternFromGlobs(const std::string & initial_str_with_glob
             ReadBufferFromString buf_range(buffer);
             buf_range >> range_begin >> point >> point >> range_end;
             bool leading_zeros = buffer[0] == '0';
-            size_t num_len = std::to_string(range_end).size();
+
+            size_t range_begin_width = buffer.find('.');
+            size_t range_end_width = buffer.size() - buffer.find_last_of('.') - 1;
+
+            size_t num_len = 0;
```
Comment (rschu1ze) [inline src/Common/parseGlobs.cpp:58]: Maybe "output_width" instead of "num_len"?

File: src/Common/parseGlobs.cpp (hunk 2 above)
Comment (rschu1ze) [reply src/Common/parseGlobs.cpp:62]: Suggest to remove the comment here (i.e. only retain the comment in l. 60). At it confused me more than helps.

File: src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp [hunk 5]
```diff
@@ -8,14 +8,29 @@ using namespace DB;
 
 TEST(Common, makeRegexpPatternFromGlobs)
 {
+
```
Comment (rschu1ze) [inline src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp:11]: Thanks for adding the new tests. To make this even more systematic and just as a suggestion: We could add "sub-cases" for "case 1" and "case 2" where 1) the left range width == the right range width, 2) then examples where the left range width > the right range width and 3) finally examples where the left range width < the right range width. Sub-case 2) can only be achieved using leading zeros on the left side, which means we might merge case 3 into cases 1 and 2.)

File: src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp [hunk 6]
```diff
@@ -8,14 +8,29 @@ using namespace DB;
...
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{10..20}"), "f(10|11|12|13|14|15|16|17|18|19|20)");
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{100..90}"), "f(90|91|92|93|94|95|96|97|98|99|100)");
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{99..109}"), "f(99|100|101|102|103|104|105|106|107|108|109)");
+    //Case 2: Descending Sequences
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{20..15}"), "f(15|16|17|18|19|20)");
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{200..199}"), "f(199|200)");
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{103..95}"), "f(95|96|97|98|99|100|101|102|103)");
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{95..103}"), "f(95|96|97|98|99|100|101|102|103)");
+    //Case 3: Sequences Start with 0's
+    EXPECT_EQ(makeRegexpPatternFromGlobs("f{00..10}"), "f(00|01|02|03|04|05|06|07|08|09|10)");
```
[the remark this answers is not in the document]
Comment (HeenaBansal2009) [reply src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp:28]: echo {00..10} returns 00|01|02|03|04|05|06|07|08|09|10 in shell and the same is with current implementation. 
Is this not hat we expect?
```
heena@Clickhouse1:~$ echo {00..10}
00 01 02 03 04 05 06 07 08 09 10
heena@Clickhouse1:~$ 

```

In clickhouse Doc as well , this is what we expect from {000..999}
image

Please advice.

**Comment by alexey-milovidov** [comment]
Please make it work exactly as
```
$ echo {0..10}
0 1 2 3 4 5 6 7 8 9 10
$ echo {00..10}
00 01 02 03 04 05 06 07 08 09 10
$ echo {001..10}
001 002 003 004 005 006 007 008 009 010
```

**Comment by HeenaBansal2009** [comment]
> Please make it work exactly as
> 
> ```
> $ echo {0..10}
> 0 1 2 3 4 5 6 7 8 9 10
> $ echo {00..10}
> 00 01 02 03 04 05 06 07 08 09 10
> $ echo {001..10}
> 001 002 003 004 005 006 007 008 009 010
> ```

Yes, It is working as above now.

File: src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp:28]: Bash and ksh expand `{01..10}` to `0 1 2 3 4 5 6 7 8 9 10` while zsh expands to `01 02 03 04 05 06 07 08 09 10` :dizzy_face: Guess we should follow the more straightforward style which is (imho) zsh. We are good then.

**Comment by rschu1ze** [comment]
There are still failures in the new tests which is strange given that the tests are deterministic. Maybe merge from master to trigger a new build/tests?

File: tests/queries/0_stateless/02297_regex_parsing_file_names.sh [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+${CLICKHOUSE_CLIENT} -q "DROP TABLE IF EXISTS t_regex"
+
+${CLICKHOUSE_CLIENT} -q "CREATE TABLE t_regex (id UInt64, obj String, s String) ENGINE =File(JSONEachRow)" ;
+
+
+${CLICKHOUSE_CLIENT} -q "INSERT INTO t_regex SELECT * FROM file('file_{0..10}.json','JSONEachRow')";
+${CLICKHOUSE_CLIENT} -q "SELECT count() from t_regex"
+
+rm -rf "$CURDIR"/user_files;
```
[the remark this answers is not in the document]
Comment (HeenaBansal2009) [reply tests/queries/0_stateless/02297_regex_parsing_file_names.sh:37]: ooh Yes ,It was overlooked. Thanks for pointing out.

**Comment by HeenaBansal2009** [comment]
> > The new test FT testcase added in this PR was failing, I fixed It .
> 
> Just ran the unit test on the four commits in this PR locally and it failed:
> 
> ```
> ~/repos/ClickHouse2/build/src (regex =) $ ./unit_tests_dbms --gtest_filter=*makeRegexpPatternFromGlobs*
> Note: Google Test filter = *makeRegexpPatternFromGlobs*
> [==========] Running 1 test from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 1 test from Common
> [ RUN      ] Common.makeRegexpPatternFromGlobs
> /home/ubuntu/repos/ClickHouse2/src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp:39: Failure
> Expected equality of these values:
>   makeRegexpPatternFromGlobs("f{9..01}")
>     Which is: "f(1|2|3|4|5|6|7|8|9)"
>   "f(01|02|03|04|05|06|07|08|09)"
> /home/ubuntu/repos/ClickHouse2/src/Common/tests/gtest_makeRegexpPatternFromGlobs.cpp:40: Failure
> Expected equality of these values:
>   makeRegexpPatternFromGlobs("f{9..000}")
>     Which is: "f(0|1|2|3|4|5|6|7|8|9)"
>   "f(000|001|002|003|004|005|006|007|008|009)"
> [  FAILED  ] Common.makeRegexpPatternFromGlobs (4 ms)
> [----------] 1 test from Common (4 ms total)
> 
> [----------] Global test environment tear-down
> [==========] 1 test from 1 test suite ran. (4 ms total)
> [  PASSED  ] 0 tests.
> [  FAILED  ] 1 test, listed below:
> [  FAILED  ] Common.makeRegexpPatternFromGlobs
> ```

I fixed the code. When range_begin> range_end , leading zero was getting calculated based on range_end.

**Comment by HeenaBansal2009** [comment]
> The unit test runs fine now.
> 
> The integration test [0] runs "02297_regex_parsing_file_names.sh" 100 times and many times it fails. The log files [1] show:
> 
> ```
> 2022.07.06 07:35:28.862499 [ 919 ] {2533a9f7-126f-4809-9f1b-dc7abe2bcd6c} <Error> executeQuery: Code: 400. DB::ErrnoException: Cannot stat file /var/lib/clickhouse/user_files/file_2.json, errno: 2, strerror: No such file or directory: While executing File. (CANNOT_STAT) (version 22.7.1.1) (from [::1]:33596) (comment: 02297_regex_parsing_file_names.sh) (in query: INSERT INTO t_regex SELECT * FROM file('file_{0..10}.json','JSONEachRow')), Stack trace (when copying this message, always include the lines below):
> ```
> 
> Also, the SELECT sometimes returns wrong results, e.g. multiples of 11 (= the expected result), e.g. 22 and 88.
> 
> So it seems that the JSON files and the SQL table are not properly deleted between runs? I don't have a good explanation for that as both are cleaned up at the end of the test. Maybe you have an idea?
> 
> [0] https://s3.amazonaws.com/clickhouse-test-reports/38502/d89ba2e5d98600e1bf682b987339f132b8a6b6cb/stateless_tests_flaky_check__address__actions_.html [1] https://s3.amazonaws.com/clickhouse-test-reports/38502/d89ba2e5d98600e1bf682b987339f132b8a6b6cb/stateless_tests_flaky_check__address__actions_/clickhouse-server.err.log

@rschu1ze , hmm.... The tables and files are getting deleted at the end of test. Interesting .. Let me check.
Does file engine reads data from somewhere else too ?

**Comment by rschu1ze** [comment]
Maybe try to run the integration test in a loop locally?

> Does file engine reads data from somewhere else too ?

Not that I know of. I don't really have an idea what went wrong :-(

**Comment by HeenaBansal2009** [comment]
> @HeenaBansal2009 Since the error persists, respectively the last commit totally broke the test: You might try the following. I ran it locally for 30 min in a loop w/o error.
> 
> ```
> #!/usr/bin/env bash
> # Tags: no-parallel
> 
> CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
> # shellcheck source=../shell_config.sh
> . "$CURDIR"/../shell_config.sh
> 
> # Data preparation.
> 
> # Now we can get the user_files_path by use the table file function for trick. also we can get it by query as:
> #  "insert into function file('exist.txt', 'CSV', 'val1 char') values ('aaaa'); select _path from file('exist.txt', 'CSV', 'val1 char')"
> CLICKHOUSE_USER_FILES_PATH=$(clickhouse-client --query "select _path, _file from file('nonexist.txt', 'CSV', 'val1 char')" 2>&1 | grep Exception | awk '{gsub("/nonexist.txt","",$9);
> > print $9}')
> 
> mkdir -p ${CLICKHOUSE_USER_FILES_PATH}/
> 
> # rm -rf ${CLICKHOUSE_USER_FILES_PATH}/file_{0..10}.json
> rm -rf ${CLICKHOUSE_USER_FILES_PATH}/file_{0..10}.csv
> 
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_0.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_1.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_2.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_3.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_4.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_5.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_6.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_7.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_8.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_9.csv
> echo '0' > ${CLICKHOUSE_USER_FILES_PATH}/file_10.csv
> 
> # echo '' > ${CLICKHOUSE_USER_FILES_PATH}/file_10.csv
> 
> ${CLICKHOUSE_CLIENT} -q "DROP TABLE IF EXISTS t_regex;"
> 
> ${CLICKHOUSE_CLIENT} -q "CREATE TABLE t_regex (id UInt64) ENGINE = MergeTree() order by id;"
> 
> ${CLICKHOUSE_CLIENT} -q "INSERT INTO t_regex SELECT * FROM file('file_{0..10}.csv','CSV');"
> ${CLICKHOUSE_CLIENT} -q "SELECT count() from t_regex;"
> 
> rm -rf ${CLICKHOUSE_USER_FILES_PATH}/file_{0..10}.csv;
> ${CLICKHOUSE_CLIENT} -q "DROP TABLE IF EXISTS t_regex;"
> ```
> 
> If that also fails (from time to time) we know at least that the problem isn't related to the test itself. I'd also be fine then with deleting the integration test altogether, after all we have good unit test coverage.

Thanks @rschu1ze  for looking in to this. I'll try to update the test and see how it behaves with CSV files.

**Comment by HeenaBansal2009** [comment]
@rschu1ze , The FT testcase failing is different from the PR and after changing the file type from JSON to CSV, 
the FT testcase is no more flaky and I can see everything good.

**Comment by rschu1ze** [comment]
Good, I'll merge for now and follow up with the JSON issue (though it is not clear yet if it was a test problem).

**Pull request merged**